Proposal: Add configuration profiles to support multiple teams

flosell opened the pull request:
This PR is a proposal how #53 could be resolved. It adds the `profile` argument to the URL, allowing users to switch between as many sets of configurations as they like. 

Notes for the review:
* This PR removes two optimisations: 
  * Previously, settings-information was kept in memory. This PR changes this to go back to the DB for every request so makes performance/load a bit worse. I don't think this is a big deal for the kind of workload we are looking at (few, long-running clients) but if you have doubts and feel it's worth the extra complexity, let me know and we can add some caching
  * Previously, disabled pipelines were not fetched from GoCD. I can see how this can reduce load on GoCD if there are a lot of pipelines. However, I'd guess in most cases, someone would be interested in most of those pipelines so they'd need to be fetched anyway. But if you feel there is enough of a case to keep this optimisation, I could think of a solution where we aggregate the disabled pipelines in all the settings and then filter for those.

karmats commented:
I like this approach, but the disabled pipelines needs to not be fetched if they are disabled. I'm ok if it only happens when profile is not defined i.e. `if (profile) => fetch all, else => filter disabled`. I agree that the extra database calls is ok.

It would of course be nicer if the url was something like `/team1` instead of `?profile=team1` but since the monitor supports pipeline groups I guess it will be tough to implement that since the monitor then needs to support both `/team1/pipeline-group` and just `/pipeline-group` when profiles are not setup.

flosell commented:
Thanks for the feedback! I added some basic filtering for the pipelines being fetched based on what you proposed. The current switch is based on the database: If the database contains profiles, filtering is off.
I played around with a more advanced approach, filtering pipelines for whichever profiles had active clients but found it a bit too complex to reason about. If you are interested, have a look [here](https://github.com/flosell/gocd-monitor/commit/133f1c8947df4732ec6b1334b1f17cca999502a9)

I also changed the behaviour for new profiles (i.e. those that don't have settings stored in the database): They are now initialised with the settings stored for the default profile. This to me seems to be a bit more intuitive (the default profile configuration is something that's visible to users in the browser, the config file or environment variable isn't) and reduces the chance of edge cases (e.g. when pipelines are disabled in the stored configuration for the default profile but not in the environment-variable/config file - should we fetch those pipelines?)

Regarding the URL structure, I agree adding the teams/profiles to the URL might be a bit tricky/add more complexity.

karmats commented on server/services/GoService.js:
```diff
@@ -35,40 +36,66 @@ export default class GoService {
     this.pollGoServer();
   }
 
+  currentSettings(profile) {
```
Nice improvement 👍

karmats commented on server/services/GoService.js:
```diff
@@ -5,6 +5,7 @@ import GoTestService from './GoTestService';
 import DBService from './DBService';
 import Logger from '../utils/Logger';
 import CucumberParser from '../utils/CucumberParser';
+import _ from 'lodash'
```
Could you add this to `package.json` as well? Now it's in `node_modules` because other dependencies rely on it. Or since you just use it for just one function, could you write in in plain javascript? I like to keep dependencies as few as possible

flosell replied:
Thanks for the feedback! Turns out a solution without lodash is actually simpler :) Fixed.

karmats commented on spec/GoServiceSpec.js:
```diff
@@ -29,14 +31,14 @@ describe('GoService spec', () => {
       expect(goService.clients).to.have.lengthOf(0);
 
-      goService.registerClient({ id: 'client1', on: () => { }, emit: () => { } });
+      goService.registerClient(clientStub());
```
🌟 Great that the tests gets some love as well.
